### PR TITLE
[WebSocket] Refactor websocket module.

### PR DESCRIPTION
**Pull request by Faless**
This commit is a huge refactor of the websocket module. The module is really old, and some design choices had to be re-evaluated.

The WebSocketClient and WebSocketServer classes are now gone, and WebSocketPeer can act as either client or server.
The WebSocketMultiplayerPeer class is no longer abstract, and implements the Multiplayer API on top of the lower level WebSocketPeer.

WebSocketPeer is now a "raw" peer, like StreamPeerTCP and StreamPeerTLS, so it emits no signal, and just needs polling to update its internal state.

To use it as a client, simply call WebSocketPeer.coonect_to_url, then frequently poll the peer until STATE_OPEN is reached and then you can write or read from it, or STATE_CLOSED and then you can check the disconnect code and reason).

To implement a server instead, a TCPServer must be created, and the accepted connections needs to be provided to
WebSocketPeer.accept_stream (which will perform the HTTP handshake).

A full example of a WebSocketServer using TLS will be provided in the demo repository.

Draft status:
- [x] ~Needs a PR to update chat and multiplayer demo.~ https://github.com/godotengine/godot-demo-projects/pull/781
- [X] ~Decide if we still want to provide an official "WebSocketServer" class, either in the module, or (I think better), in GDScript.~ https://github.com/godotengine/godot-demo-projects/pull/781 also has a reference WebSocketClient and WebSocketServer GDScript implementation.
- [x] ~Evaluate exposing `send(String text)` as `send(Variant message, force_write_mode)` accepting either `String` or `PackedByteArray`.~ Exposed as `send(message : PackedByteArray, p_write_mode := WRITE_MODE_BINARY)` and `send_text(message: String)`.
- [x] ~Expose `accept_stream(stream, do_handshake=true)`~. This feels like bloat, since it would also require special methods for the multiplayer peer, so I was thinking we might add a `handshake_callback` instead, to perform custom handshake, but I'd rather do that in a separate PR.

I sadly didn't manage to finish this for the beta freeze, nevertheless I'd rather provide a GDScript compatibility layer and still merge this in 4.0, since it's a great improvement in both API design (now in line with other peers) and potential future development like better buffers management and full user control over the handshake process on supported platforms..

### CODE REVIEW

**Comment by Faless** [comment]
Ready for review

**Review by dsnopek — COMMENTED** [review]
At a high-level, I think this is a great refactor! Working with WebSockets has always been a little awkward, mainly to do with getting the peer via `WebSocketClient.get_peer(1)` in order to actually send and receive packets. The first time I worked with it, I remember it taking me a little time to get my head around that.

I converted one real use of `WebSocketClient` to `WebSocketPeer` from this PR, and it was very straight-forward, except for figuring out how to handle a close request, which I think should be mentioned in the docs as I described in my other comments. There's one more real use of `WebSocketClient` that I want to try converting - I'll report back after, but I suspect it'll go fine as well :-)

Great work! :-)

**Comment by dsnopek** [comment]
I converted the other real use of WebSockets I mentioned above. Both of the conversions I did were pretty easy, because they both had a dedicated class for using the WebSocket. I could certainly imagine that converting code that just had a `WebSocketClient` mixed in with many other things could be trickier and less user-friendly - for that case, the GDScript compatibility layer that @Faless mentioned would make things much more user-friendly for developers. I guess it's just a question of how to share that code?

File: modules/websocket/doc_classes/WebSocketPeer.xml [hunk 1]
```diff
@@ -74,12 +141,42 @@
...
 			Specifies that WebSockets messages should be transferred as binary payload (any byte combination is allowed).
 		</constant>
+		<constant name="STATE_CONNECTING" value="0" enum="State">
+			Socket has been created. The connection is not yet open.
+		</constant>
+		<constant name="STATE_OPEN" value="1" enum="State">
+			The connection is open and ready to communicate.
+		</constant>
+		<constant name="STATE_CLOSING" value="2" enum="State">
+			The connection is in the process of closing.
```
[the remark this answers is not in the document]
Comment (Faless) [reply modules/websocket/doc_classes/WebSocketPeer.xml]: State closing usually means we have sent a close request, but haven't received confirmation yet.
So, it isn't really safe to call `get_close_code` and `get_close_reason` here, but should be called when reaching `STATE_CLOSED`. I've clarified that in the docs.

File: modules/websocket/doc_classes/WebSocketPeer.xml [hunk 2]
```diff
@@ -1,25 +1,72 @@
...
+		extends Node
+
+		var socket = WebSocketPeer.new()
+
+		func _ready():
+		    socket.connect_to_url("wss://example.com")
+
+		func _process(delta):
+		    socket.poll()
+		    while socket.get_ready_state() == STATE_OPEN and socket.get_available_packet_count():
```
[the remark this answers is not in the document]
Comment (Faless) [reply modules/websocket/doc_classes/WebSocketPeer.xml]: Agreed, I added a `pass` and a comment for `WebSocketPeer.STATE_CLOSING` (in that state, you only have to wait and keep polling), and a print explaining the close states for `WebSocketPeer.STATE_CLOSED`.

**Comment by Faless** [comment]
> I guess it's just a question of how to share that code?

That's a bit tricky indeed, the best way is probably as an addon in the asset library, I think the [`WebSocketClient`](https://github.com/godotengine/godot-demo-projects/pull/781/files#diff-8c937275f16b496979b43d1d524581c03eaa550d0037aa349b4e53a31b803fe1) and [`WebSocketServer`](https://github.com/godotengine/godot-demo-projects/pull/781/files#diff-95b4998a5c847e7721ea7c64c0e0e36a1efbc3e7e377332c853dc82ae1ca9645) in the demo can act as a basis, they are not 1-to-1 to the old ones, but I think are a good "signal-based" example that should be easily adapted.

File: modules/websocket/doc_classes/WebSocketMultiplayerPeer.xml [hunk 3]
```diff
@@ -10,34 +10,80 @@
...
 	</tutorials>
 	<methods>
+		<method name="close">
+			<return type="void" />
+			<description>
+				Closes this [MultiplayerPeer], resetting the state to [constant MultiplayerPeer.CONNECTION_CONNECTED].
+				[b]Note:[/b] To make sure remote peers receive a clean close prefer disconnecting clients via [method disconnect_peer].
+			</description>
+		</method>
+		<method name="create_client">
```
Comment (akien-mga) [inline modules/websocket/doc_classes/WebSocketMultiplayerPeer.xml:20]: Missing doc for this one.

**Review by akien-mga — APPROVED** [review]
Approved in PR review meeting.

**Comment by Faless** [comment]
I've temporarily copied the GDScript nodes to https://github.com/Faless/gd-websocket-nodes and structured it as an addon.
Still needs some work (proper screenshot, better description) and would be happy to move it to an "official extensions repository" once we have one if there is interest.

**Comment by Anutrix** [comment]
The docs for WebSocket badly need an update.

**Pull request merged**